docs/design-proposals/node/Unified-Vertical-Scaling-for-Container-and-VM-Pods.md

```diff
@@ -0,0 +1,646 @@
+  to the ResourcesAllocated field,
+* new admission controller to limit update access to ResourcesAllocated field
+  only to Node, and mutates any updates to ResourcesAllocated & ResizePolicy
+  fields to maintain compatibility with older versions of clients,
+* added validation allowing only CPU and memory resource changes,
+* setting defaults for ResourcesAllocated and ResizePolicy fields.
+
+Admission Controllers: LimitRanger, ResourceQuota need to support Pod Updates:
+* for ResourceQuota, podEvaluator.Handler implementation is modified to allow
+  Pod updates, and verify that sum of Pod.Spec.Containers[i].Resources for all
+  Pods in the Namespace don't exceed quota,
+* PodResourceAllocation admission plugin is ordered before ResourceQuota.
+* for LimitRanger we check that a resize request does not violate the min and
+  max limits specified in LimitRange for the Pod's namespace.
+
+Kubelet:
+* set Pod's Status.ContainerStatuses[i].Resources for Containers upon placing
+  a new Pod on the Node,
+* update Pod's Spec.Containers[i].ResourcesAllocated upon resize,
+* change UpdateContainerResources CRI API to work for both Linux & Windows.
+
+Scheduler:
+* compute resource allocations using Pod.Spec.Containers[i].ResourcesAllocated.
+
+Other components:
+* check how the change of meaning of resource requests influence other
+  Kubernetes components.
+
+### Changes in Arktos-vm-runtime-service
+
+#### Background and context
+Currently in Arktos, the VM workload POD is defined the same way as container pod in terms of resources in the spec, i.e.
+
+    spec:
+      resource:
+       limits:
+         cpu:
+         memory:
+       request:
+         cpu:
+         memory:
+     
+Arktos controls the resource the same way as for containers, i.e. schedule based on requested resource level, and evict 
+POD based on the limits values, per some defined eviction policies.
+
+Unlike the container, which is essentially the isolated process on the host, there are quite some differences for VM that
+are worth to describe here to lay some common ground before the design for supporting vertical scaling for VMs.
+
+For VM types, the resources for VM are also controlled by the Hypervisor and the management layer i.e. the Libvirt and the 
+Qemu/KVM/Xen. For memory resource allocation, a few essential configurations are in the Libvirt domain definition:
+
+    <domain>
+      ...
+      <maxMemory slots='16' unit='KiB'>1524288</maxMemory>
+      <memory unit='KiB'>524288</memory>
+      <currentMemory unit='KiB'>524288</currentMemory>
+      ...
+    </domain>
+
+Memory defines the max memory allocation at the boot time, while the currentMemory defines the real memory allocation for
+the guest OS. The delta between the currentMemory and memory is the space for memory ballooning to "reclaim" memory
+between VMs and the hypervisor. The maxMemory defines the max memory allocation for a VM at runtime, which can be achieved
```

Comment:
Since currentMemory is the real memory allocated to VM, how can ballooning reclaim memory from delta (i.e. the memory amount beyond currentMemory)? My understanding is (0, currentMemory) is the space that ballooning can reclaim.

Reply:
yes, i should describe it clearly. This paragraph is mainly to describe the upper boundary of ballooning can achieve in term of scale-up the memory of the VM. to scale-down, it is the "least memory" of the guestOS ( say 512 Mi ) up to currentMemory allocation. i will clarify.